Optimise builds

tobiasge opened the pull request:
## New Behavior

- More builds are done in parallel
- Builds are only run when necessary (changes in Netbox, Netboc-docker, Python image)

## Contrast to Current Behavior

- See above

## Discussion: Benefits and Drawbacks
When following the "latest" or "v2.7" tags in Kubernetes or Openshift the containers were deployed everytime the GH Action ran. From now on the images are build only when something changed. So the automatic deployment is only triggered when really needed. 
The image count was also reduced. Changes to the master branch in Netbox are only made when a new release is ready. So we don't need to build extra images for the latest tag. This tag now follows the newest version of Netbox.

## Changes to the Wiki

- "Build" page: Adapt the list of build scripts

## Proposed Release Note Entry

- Maintenance: Improved build system

## Double Check
* [x] I have read the comments and followed the PR template.
* [x] I have explained my PR according to the information in the comments.
* [x] My PR targets the `develop` branch.

tobiasge committed:
Add checks to verifiy if a new build is needed

This checks if the source materials (python image, Netbox commit,
netbox-docker commit) have changed since the last build. This check is done
by comparing the digest and commit ids from the previous image with the
given tag to the current values taken from the Git and Docker repositories.

The checks are only performed for builds by the automated builds on Github.

tobiasge committed:
Fix shellcheck items found by shellcheck 0.7.1

tobiasge committed:
Changed "build-branches.sh" to "build-next.sh"

The old version of "build-branches.sh" skipped the pushing of images when one of
them didn't change its sources. When looking at the Netbox branches I noticed
that the "master" branch only changes when there is a new release. Because we
build the new releases anyway in "build-latest.sh" that leaves "develop" and
"develop-*" which change regularly. The new script "build-next.sh" is responsible
for building "develop-*" as the next major release of Netbox. The build of
"develop" is moved to the Github Action build matrix. This has the additional
advantage of being faster because more builds are done in parallel.

tobiasge committed:
Let "latest" tag follow the highest version

Latest tag is now added to the new version when it is release by Netbox.

tobiasge committed:
Removed Travis CI configuration and script

We are no longer building the images with Travis CI.

tobiasge committed:
Fixed "latest" tag

The variable for the latest tag didn't contain all the needed values.

tobiasge committed:
Fixes build when DOCKER_FROM is set but empty

The DOCKER_FROM is set to an empty value in the push tests.
This expands the check to catch this test case

cimnine commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-ARG FROM=python:3.7-alpine
+ARG FROM
```
Why did you remove the default?

tobiasge replied:
To check if the source image changed, we have to know what it is. So I moved the default to the `build.sh` script.

cimnine commented on build.sh:
```diff
@@ -217,22 +224,68 @@ for DOCKER_TARGET in "${DOCKER_TARGETS[@]}"; do
+      SHOULD_BUILD="true"
+      BUILD_REASON="${BUILD_REASON} interactive"
+    fi
+    if [ "$DOCKER_REGISTRY" = "docker.io" ] && [ "$SHOULD_BUILD" = "false" ]; then
```
```suggestion
    elsif [ "$DOCKER_REGISTRY" = "docker.io" ]; then
```

cimnine commented:
When I read through this, I start to think we should rewrite the build script in something else than bash, Python or so 🙈

tobiasge commented:
I think for now Bash is OK, but if we want to add more functionality we should do that.

tobiasge committed:
Fixed script in case no branch for next exists

Build now prints a message when no devlop-* branch is found instead of crashing

tobiasge commented:
The builds in my [repository](https://github.com/tobiasge/netbox-docker/actions) with the new script have been stable for a while now. I think we can merge this.

cimnine commented on build.sh:
```diff
@@ -269,6 +323,10 @@ for DOCKER_TARGET in "${DOCKER_TARGETS[@]}"; do
     fi
+    if [ -n "${BUILD_REASON}" ]; then
+      BUILD_REASON=$(sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//' <<< "$BUILD_REASON")
+      DOCKER_BUILD_ARGS+=( --label "BUILD_REASON=${BUILD_REASON}" )
```
What is your idea that you add this as a label? Wouldn't it be enough to just print it to the console?

tobiasge replied:
I thought it would be use full to know why a particular image was build without searching through the build logs.

cimnine reviewed (commented):
Did you run [shellcheck](https://www.shellcheck.net/) against shell files?

tobiasge commented:
> Did you run [shellcheck](https://www.shellcheck.net/) against shell files?

I did. But I found just found one more `shellcheck` problem. Will commit a fix to this branch.

tobiasge committed:
Update build.sh

Include suggestion from cimnine

Co-authored-by: Christian Mäder <cimnine@users.noreply.github.com>

tobiasge committed:
Fix shellcheck error

cimnine reviewed (approved):
I still don't understand the whole file, but since it seems working in your branch, let's go for it.

cimnine commented:
I believe we must update the repository settings shortly after this is merged to specify which of the new jobs are required to run and pass in order for a commit to get 'green'.
